common/ComputationalTools.h

```diff
@@ -0,0 +1,40 @@
+#ifndef COMPUTATIONAL_TOOLS
+#define COMPUTATIONAL_TOOLS
+
+template <typename T, typename Total, int N>
+class MovingAverage
```

Comment:
A separate type for the total and the samples is a little weird. You could just combine them.
Also, maybe have a restriction on N being a power of two, so then the modulus in the take sample function can be optimized to a bitwise and.